mantidproject/mantid
# Make handling of expected peak units more robust in EnggFitPeaks

Previously, if passed a list of peak centres in TOF, `EnggFitPeaks` assumed they were in dSpacing, and so converted them to TOF as if they were, leading to incorrect expected peak centres to be passed to the fitting routines.

This PR makes the handling of units in EnggFitPeaks more robust. If none of the expected peaks lie in the limits of the x axis of the input workspace, then EnggFitPeaks converts them from TOF to D, and sees if any of the converted peaks lie within the limits. If not, the algorithm fails.

This change enables the peak picker to be used in the ED GUI, as peaks are normally selected from a plot in TOF.

**To test:**

Make sure EnggFitPeaks unit test passes locally and remotely.

Make a sample workspace:
```

peak_def1 = "name=FlatBackground,A0=1;" \
            "name=BackToBackExponential, I=15000, A=0.1, B=0.14, X0=41209, S=50"
peak_def2 = "name=FlatBackground,A0=1; " \
            "name=BackToBackExponential, I=6000, A=0.02, B=0.021, X0=46895, S=40"

sample_ws = CreateSampleWorkspace(Function="User Defined",
                                  UserDefinedFunction=peak_def1 + ";" + peak_def2,
                                  NumBanks=1, BankPixelWidth=1, XMin=40000, XMax=50000, 
                                  BinWidth=25)

EditInstrumentGeometry(Workspace=sample_ws, L2=[1.5], Polar=[90], PrimaryFlightPath=50)
```

First try running EnggFitPeaks with the expected peak centres in TOF:
```
fit_results = EnggFitPeaks(InputWorkspace=sample_ws, WorkspaceIndex=0, 
                           ExpectedPeaks=[41209, 46895])
```

The fit should run successfully. The `dSpacing` column in `fit_results` should contain the equivalent peak centres - approximately 2.238 and 2.547.

Running the fit with the numbers:
```
fit_results = EnggFitPeaks(InputWorkspace=sample_ws, WorkspaceIndex=0, 
                           ExpectedPeaks=[2.238, 2.547])
```
Should give you a similar table.

Finally, EnggFitPeaks supports a GSAS-style conversion from TOF to D if the relevant parameters are set on the workspace. Try the following:
```
run = sample_ws.mutableRun()
run.addProperty("difa", 0.0, False)
run.addProperty("difc", 18400.0, False)
run.addProperty("tzero", 4.0, False)

fit_results = EnggFitPeaks(InputWorkspace=sample_ws, WorkspaceIndex=0, 
                           ExpectedPeaks=[41209, 46895])
```
And verify, again, that the `dSpacing` column contains sensible values (in this case approx 2.239 and 2.548).

Fixes #21342 

**Release Notes** 
[Here](https://github.com/mantidproject/mantid/pull/21355/commits/a49d6c938cba484b0e367186151f8c69fe867a64)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
